internal/controller/helpers.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/projectcontour/contour/internal/k8s"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi_v1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// setRouteCondition sets a condition on a Gateway API route
+// for a given gateway & controller. It performs an upsert,
+// updating the existing condition (if there has been a change),
+// otherwise adding a new condition.
+func setRouteCondition(
+	routeParentStatuses []gatewayapi_v1alpha2.RouteParentStatus,
+	condition metav1.Condition,
+	gateway *gatewayapi_v1alpha2.Gateway,
+	controllerName gatewayapi_v1alpha2.GatewayController,
+) []gatewayapi_v1alpha2.RouteParentStatus {
+	// Look for a RouteParentStatus for the relevant Gateway.
+	for _, parentStatus := range routeParentStatuses {
+		if !isRefToGateway(parentStatus.ParentRef, k8s.NamespacedNameOf(gateway)) {
+			continue
+		}
+
+		for i := range parentStatus.Conditions {
+			cond := &parentStatus.Conditions[i]
+
+			if cond.Type != condition.Type {
+				continue
+			}
+
+			// Update only if something has changed.
+			if cond.Status != condition.Status || cond.Reason != condition.Reason || cond.Message != condition.Message {
+				cond.Status = condition.Status
+				cond.Reason = condition.Reason
+				cond.Message = condition.Message
+				cond.LastTransitionTime = condition.LastTransitionTime
+				cond.ObservedGeneration = condition.ObservedGeneration
+			}
+
+			return routeParentStatuses
+		}
+
+		// condition of type "Accepted" does not already exist for the parent:
+		// append it.
+		parentStatus.Conditions = append(parentStatus.Conditions, condition)
+		return routeParentStatuses
+	}
+
+	// RouteParentStatus for the Gateway does not already exist: add it.
+	routeParentStatuses = append(routeParentStatuses, gatewayapi_v1alpha2.RouteParentStatus{
+		ParentRef:      gatewayapi.GatewayParentRef(gateway.Namespace, gateway.Name),
+		ControllerName: controllerName,
+		Conditions:     []metav1.Condition{condition},
+	})
+
+	return routeParentStatuses
+}
+
+// isRefToGateway returns whether the given ref refers to a Gateway with the
+// specified name and namespace.
+func isRefToGateway(ref gatewayapi_v1alpha2.ParentRef, gateway types.NamespacedName) bool {
+	return ref.Group != nil && *ref.Group == gatewayapi_v1alpha2.GroupName &&
+		ref.Kind != nil && *ref.Kind == "Gateway" &&
+		ref.Namespace != nil && *ref.Namespace == gatewayapi_v1alpha2.Namespace(gateway.Namespace) &&
+		string(ref.Name) == gateway.Name
+}
+
+// referencesContourGateway returns whether a given list of
+// ParentRefs references a Gateway controlled by this Contour.
+func referencesContourGateway(
+	parentRefs []gatewayapi_v1alpha2.ParentRef,
+	routeNamespace string,
+	controllerName gatewayapi_v1alpha2.GatewayController,
+	kubeClient client.Client,
+	log logrus.FieldLogger,
+) (bool, *gatewayapi_v1alpha2.Gateway) {
```

Comment:
Nit: Feels like it would be more idiomatic to have this the other way round, so that we can use it like the other two value functions that return an `ok` variable?
```suggestion
) (b*gatewayapi_v1alpha2.Gateway, bool) {
```
would allow the caller to do:
```go
gw, ok := referencesContourGateway(...)
if !ok {
}
```

Reply:
:+1: done